locator/token_metadata.cc

```diff
@@ -763,122 +713,108 @@ token_metadata_impl::has_pending_ranges(sstring keyspace_name, inet_address endp
-        all_left_metadata->update_normal_tokens(tokens, endpoint).get();
-        auto address_ranges = strategy.get_ranges(endpoint, *all_left_metadata).get0();
-        for (const dht::token_range& x : address_ranges) {
-            new_pending_ranges.emplace(x, endpoint);
-        }
-        all_left_metadata->_impl->remove_endpoint(endpoint);
-    }
-    all_left_metadata->_impl->sort_tokens();
-}
-
 future<> token_metadata_impl::update_pending_ranges(
         const token_metadata& unpimplified_this,
         const abstract_replication_strategy& strategy, const sstring& keyspace_name, dc_rack_fn& get_dc_rack) {
     tlogger.debug("calculate_pending_ranges: keyspace_name={}, bootstrap_tokens={}, leaving nodes={}, replacing_endpoints={}",
         keyspace_name, _bootstrap_tokens, _leaving_endpoints, _replacing_endpoints);
     if (_bootstrap_tokens.empty() && _leaving_endpoints.empty() && _replacing_endpoints.empty()) {
         tlogger.debug("No bootstrapping, leaving nodes, replacing nodes -> empty pending ranges for {}", keyspace_name);
-        set_pending_ranges(keyspace_name, std::unordered_multimap<range<token>, inet_address>(), can_yield::no);
+        _pending_ranges_interval_map.erase(keyspace_name);
         return make_ready_future<>();
     }
 
     return async([this, &unpimplified_this, &strategy, keyspace_name, &get_dc_rack] () mutable {
-        std::unordered_multimap<range<token>, inet_address> new_pending_ranges;
-        calculate_pending_ranges_for_replacing(unpimplified_this, strategy, new_pending_ranges);
-        // Copy of metadata reflecting the situation after all leave operations are finished.
-        auto all_left_metadata = make_token_metadata_ptr(clone_after_all_left().get0());
-        calculate_pending_ranges_for_leaving(unpimplified_this, strategy, new_pending_ranges, all_left_metadata);
-        // At this stage newPendingRanges has been updated according to leave operations. We can
-        // now continue the calculation by checking bootstrapping nodes.
-        calculate_pending_ranges_for_bootstrap(strategy, new_pending_ranges, all_left_metadata, get_dc_rack);
-        all_left_metadata->clear_gently().get();
-
-        // At this stage newPendingRanges has been updated according to leaving and bootstrapping nodes.
-        set_pending_ranges(keyspace_name, std::move(new_pending_ranges), can_yield::yes);
+        // new_token_metadata incorporates all the changes from leaving, bootstrapping and replacing
+        const auto new_token_metadata = token_metadata([&]() -> std::unique_ptr<token_metadata_impl> {
+            auto result = clone_only_token_map(false).get0();
+
+            // construct new_normal_tokens based on _bootstrap_tokens and _replacing_endpoints
+            std::unordered_map<inet_address, std::unordered_set<token>> new_normal_tokens;
+            if (!_replacing_endpoints.empty()) {
+                for (const auto& [token, inet_address]: _token_to_endpoint_map) {
```

Comment:
I feel like we're missing some important `maybe_yield()`s. There were a bunch of those in the old code, this code seems to have fewer of them.

To be safe, maybe let's add a `maybe_yield()` whenever we're iterating over tokens: here, below for `_bootstrap_tokens` loop, then `new_normal_tokens` loop.

Curious that `sort_tokens()` doesn't yield though. Maybe it's not necessary. The loops are almost linear (with `log` here and there). Still, the yield shouldn't hurt (cc @bhalevy - please comment)

Reply:
I was also curios about those `maybe_yield`s:
  * why we have so many of them in the `token_metadata` code? Most of the loops are linear, number of elements should be small, they all should be taking nanoseconds to execute.
  * what is the cost of them? `need_preempt` is `atomic_signal_fence(std::memory_order_seq_cst)` and two relaxed reads, I'm not sure if this heavier than the iterations of the loops.

Reply:
> why we have so many of them in the token_metadata code? Most of the loops are linear, number of elements should be small, they all should be taking like nanoseconds to execute.

Probably because these smaller functions are called from larger functions which sometimes have quadratic complexity etc. These yields were being added over time, solving reactor stalls.

> what is the cost of them? need_preempt is atomic_signal_fence(std::memory_order_seq_cst) and two relaxed reads, I'm not sure if this is heavier than the iterations of the loops.

Note that `maybe_yield` doesn't yield unconditionally, hence the `maybe_`. I don't know about the cost of yield itself though. But it should be insignificant.

Reply:
> Note that maybe_yield doesn't yield unconditionally, hence the maybe_. 

Yes, but it uses `need_preempt` to decide whether or not to yield.

> Probably because these smaller functions are called from larger functions which sometimes have quadratic complexity etc.

This means we can have just one `maybe_yield` per function, not per loop iteration.

Reply:
The respective patches should provide the context for why the maybe_yield calls were added. Basically this was in response to stalls seen with a large number of tokens.
As for sorting the tokens I actually had a branch with a gentle sorting algorithm that would consider yielding in the loop instead of using std::sort.  IIRC it required a very large number of tokens to reproduce a significant stall when sorting that I thought it's not worth submitting it, but maybe we should reconsider it now that we have large enough clusters out there.

Anyhow, anything that has quadratic complexity or worse of |tokens| needs yielding. Linear should be fine, O(n*log(n)) is marginal, depnding on n and the constant multiplier.

Reply:
Besides sorting tokens, the only part of the code that is above linear is the loop to compute `migration_info`, and we already have `thread::maybe_yield` there, so no more `maybe_yield` is needed.